### ISSUE TITLE
Single Sign On fat java client unit test

### ISSUE THREAD

**Issue by ekkelenkamp** [Waffle/waffle#1194]
Hi,

First of all, thanks for the great product. Waffle is working excellent for single sign on using a tomcat web application that is AD Domain joined.
As a next step I'm investigating in how to use Waffle/JNA to get a kerberos token on the client and use it to authenticate againts our API.
I created a unit test based on the testNegotiate test method from: http://dblock.github.io/waffle/waffle-tests/xref-test/waffle/servlet/NegotiateSecurityFilterTests.html.
When trying to implement this, I didn't succeed in creating a test case, because the following part doesn't compile:
   Sspi.SecBufferDesc continueTokenBuffer = new Sspi.SecBufferDesc(Sspi.SECBUFFER_TOKEN, continueTokenBytes);
There is no public constructor  Sspi.SecBufferDesc with these arguments.
I'm using JNA 5.7.0.
I fixed this by using:

Sspi.SecBufferDesc continueTokenBuffer = new SspiUtil.ManagedSecBufferDesc(Sspi.SECBUFFER_TOKEN, continueTokenBytes);

After running the test, it fails after the second iteration.

The first token that is passed looks as follows:

Negotiate YIICHwYGKwYBBQUCoIICEzCCAg+gGjAYBgorBgEEAYI3AgIeBgorBgEEAYI3AgIKooIB7wSCAetORUdPRVhUUwAAAAAAAAAAYAAAAHAAAAA6mEjtbF4kXf0OC6ap5/MlVtrf1xpnHCP4Shm8uUDjY9q3Tc6f/eZ4Y72FkD06+10AAAAAAAAAAGAAAAABAAAAAAAAAAAAAABcM1MN6vkNTbLsSuN4bsMITkVHT0VYVFMCAAAAAQAAAEAAAAB7AQAAOphI7WxeJF39DgumqefzJVwzUw3q+Q1NsuxK43huwwhAAAAAOwEAADCCATeggfwwgfkwUYBPME0xSzBJBgNVBAMeQgBNAFMALQBPAHIAZwBhAG4AaQB6AGEAdABpAG8AbgAtAFAAMgBQAC0AQQBjAGMAZQBzAHMAIABbADIAMAAyADAAXTBRgE8wTTFLMEkGA1UEAx5CAE0AUwAtAE8AcgBnAGEAbgBpAHoAYQB0AGkAbwBuAC0AUAAyAFAALQBBAGMAYwBlAHMAcwAgAFsAMgAwADEAOQBdMFGATzBNMUswSQYDVQQDHkIATQBTAC0ATwByAGcAYQBuAGkAegBhAHQAaQBvAG4ALQBQADIAUAAtAEEAYwBjAGUAcwBzACAAWwAyADAAMQA4AF2hNjA0oBEbD1dFTExLTk9XTjpQS1UyVaEfMB2gAwIBgKEWMBQbEkRJUkVDVE9SWVxla2tlbGVuaw==

The second time:
Negotiate oUIwQKADCgEBojkEN05UTE1TU1AAAQAAAJeyCOIJAAkALgAAAAYABgAoAAAACgC6RwAAAA9MMDIyNDVESVJFQ1RPUlk=

That still results in a 401 error.
The last response only returns a list of www-authenticate headers with the values: Negotiate and NTLM, but there is no token attached.

Any thoughts on what might be the issue?

Kind Regards.

Rudie.

```java

      @Test
    public void testNegotiate() throws Exception {
        final String securityPackage = "Negotiate";
        // client credentials handle
        IWindowsCredentialsHandle clientCredentials = null;
        WindowsSecurityContextImpl clientContext = null;
        // role will contain both Everyone and SID

        try {
            // client credentials handle
            clientCredentials = WindowsCredentialsHandleImpl.getCurrent(securityPackage);
            clientCredentials.initialize();
            // initial client security context
            clientContext = new WindowsSecurityContextImpl();
            clientContext.setPrincipalName(WindowsAccountImpl.getCurrentUsername());
            clientContext.setCredentialsHandle(clientCredentials);
            clientContext.setSecurityPackage(securityPackage);
            clientContext.initialize(null, null, WindowsAccountImpl.getCurrentUsername());
            // filter chain

            System.out.println();
            while (true) {
                String clientToken = Base64.getEncoder().encodeToString(clientContext.getToken());
                HttpClient client = HttpClient.newHttpClient();
                String header =  securityPackage + " " + clientToken;
                System.out.println(header);
                HttpRequest request = HttpRequest.newBuilder()
                        .uri(new URI("http://localhost:8080/api/v1/version"))
                        .headers("Authorization", header)
                        .header("accept", "application/json")
                        .GET()
                        .build();
                HttpResponse<String> response =
                        client.send(request, HttpResponse.BodyHandlers.ofString());
                System.out.println(response.statusCode());
                System.out.println(response.headers());
                if (response.statusCode() < 300) break;
                String continueToken = response.headers().map().get("WWW-Authenticate").get(0).substring(securityPackage.length() + 1);
                byte[] continueTokenBytes = Base64.getDecoder().decode(continueToken);
                Sspi.SecBufferDesc continueTokenBuffer = new SspiUtil.ManagedSecBufferDesc(Sspi.SECBUFFER_TOKEN, continueTokenBytes);
                //Sspi.SecBufferDesc continueTokenBuffer = new Sspi.SecBufferDesc(;
                clientContext.initialize(clientContext.getHandle(), continueTokenBuffer, "localhost");
            }
        } finally {
            if (clientContext != null) {
                clientContext.dispose();
            }
            if (clientCredentials != null) {
                clientCredentials.dispose();
            }
        }
    }

```

**Comment by hazendaz** [comment]
not sure off hand but is it possible your dependency resolution is actually using older JNA?  This can happen when using spring boot and/or other solutions that use dependency management to strong arm JNA version to older revisions.  A quick run of waffle itself works with all tests at latest master so the compilation issue is most likely this situation.

**Comment by ekkelenkamp** [comment]
Actually I don't use any dependency management. Running from Intellij I can see the platform-5.7.0.jar is being used.

Looking at the breaking changes in the release notes of JNA, it suggests the api has been changed:

Since Release 5.0.0:

com.sun.jna.platform.win32.Sspi.SecBufferDesc was incompatibly changed to match the correct native semantics. SecBufferDesc describing more than one buffer were broken. For most usecases com.sun.jna.platform.win32.SspiUtil.ManagedSecBufferDesc is the best alternative.

That is why I changed the test case and used the ManagedSecBufferDesc.

Looking at the source code on github, it also shows there is only a default constructor for SecBufferDesc
https://github.com/java-native-access/jna/blob/master/contrib/platform/src/com/sun/jna/platform/win32/Sspi.java
